### ISSUE TITLE
Rewrite to proper C++

### ISSUE THREAD

**Issue by BonusPlay** [Drewol/unnamed-sdvx-clone#54]
Hey,
first of all I really enjoy the game, but I've realized the source code is completely unreadable and the project's architecture is a mess. Some things I'd like to improve on:
 - remove sources from `third_party` in favor of git submodules
 - remove unnecessary abstraction layers (for example [Audio](https://github.com/Drewol/unnamed-sdvx-clone/blob/master/Audio/include/Audio/Audio.hpp) and [Audio_Impl](https://github.com/Drewol/unnamed-sdvx-clone/blob/master/Audio/include/Audio/Audio_Impl.hpp). There is no reason to have interface if there is only one implementation).
 - removal of absolute garbage classes like [Ref](https://github.com/Drewol/unnamed-sdvx-clone/blob/master/Shared/include/Shared/Ref.hpp) in favor of STL substitutes (in this example `std::shared_ptr`).
 - refactor code to look more like C++ code and not C# code

I've started doing changes on [my fork](https://github.com/BonusPlay/unnamed-sdvx-clone), you can see most of the changes [here](https://github.com/BonusPlay/unnamed-sdvx-clone/tree/refactors). I ask for help as this is a very ambitious plan that will take very long if done by 1 man, but if done correctly it should enable more developers to help at the project.

**Comment by Drewol** [comment]
Thank you for this, it's something I've put off doing since I "took over" the project since it's mostly just a lot of tedious work without anything really changing about the game itself. I will unfortunately probably not be able to help a lot with this other than explaining some parts of the code because I'll be rather busy with studying and the work I'll still want to squeeze in is stuff like bug-fixes and smaller features.

**Comment by nashiora** [comment]
I wanted to do this when I saw the code, but I have no real experience in large C++ projects so I was hoping someone else would like to. Glad to see you've started something!

**Comment by BonusPlay** [comment]
@nashiora there are many small tasks to do, any and all help is needed.

**Comment by BonusPlay** [comment]
@nashiora yes, I also use discord a lot.

**Comment by nashiora** [comment]
@BonusPlay https://discord.gg/c6wQSG

**Comment by T045T** [comment]
@BonusPlay I saw you just got to a working build a few hours ago - nice!
Do you still need help with this transition?

**Comment by BonusPlay** [comment]
@T045T Hell yeah. Catch [this](https://trello.com/b/d2o7AmP2/game-refactor) and hack away!

**Comment by Hinara** [comment]
Just a question,
Is there any reason that USC avoid exception ?
Cause I would like to use them to improve code lisibility, but as I look in the code, everything seems to be done to avoid using them.

**Comment by BonusPlay** [comment]
@Hinara C++ exceptions provide unnecessary overhead which is aspecially undesireable in game dev (since it reduces FPS), due to compiler needing to produce stack unwinding code for exception handling. It is much better idea to use `std::expected` (which isn't out oficially, but there are ready to go implementations). You can find plenty of talks about exceptions (like [this](https://www.youtube.com/watch?v=GC4cp4U2f2E)).

**Comment by Hinara** [comment]
Yeah but some parts like map loading not done directly in a critical part, I know about exception overhead but between a program a bit less efficient, and a program which may crash easily with a raw segfault, which one is the easier to debug ?
And it makes hard to deal with some part of the stl, take std::stoi, the only alternative that comes to my mind is atoi, but atoi doesn't report any error if it fails.
And currently it's even worse because all you need is if it has failed or not, nothing about which type of error :/

**Comment by Hinara** [comment]
Optionnel seems to be a good choice for me :) I think I will use it
Doesn't no-except prevent the slowing done the code everywhere ?
```c++
int a = std::stoi("a")
```
Is that a code which is screwd up ?
Have you any solution of a standard function like stoi which return allow you to get an int but doesn't throw or send a specific integer value as an error ?
If one with optional exist I will be happy :)

**Comment by BonusPlay** [comment]
`-fno-exceptions` is quite weird. It is applied "per binarny", and you can link normal STL (with exceptions) to tour binary (without exceptions). Worst case scenario it will `abort` out.

As per stuff like `std::stoi`... C++ Game Dev never really went along with STL. People would just write their own implementation that suited their needs. I can understand that this might be a bit too much for an open source project, so I would recommend to only use exceptions as a last resort (so we can later replace them easily with something actually usable).

**Comment by Hinara** [comment]
I know this is not a real practical exemple but I've tried few things with exception -fno-exeptions and no-except
My result seems to say that an early try when it doesn't throw might be faster
https://docs.google.com/spreadsheets/d/1lFHcNHCle5fko7sSD5S-shBNEKPd3EkAI9aWDbVbfTc/edit?usp=sharing

**Comment by Hinara** [comment]
Sorry I delete my previous comment as it has some strange result which might have happened because of another load on my CPU.
So after a few tests I find out that under gnu -fno-exceptions doesn't to have that much performance benefits, however noexcept makes my code faster (benefit is not preserve with  -fno-exceptions)
Test under g++ (4:7.4.0-1ubuntu2.3)

**Comment by 123jimin** [comment]
Note that purposes of stuffs such as `Audio_Impl` are _not_ primarily focused on abstraction, but it's a technique called [PImpl](https://en.cppreference.com/w/cpp/language/pimpl), which is often used to reduce compilation time during developement.

**Comment by BonusPlay** [comment]
I'm afraid compile time is the least significant problem this project has. On my fork I started moving everything to interfaces and factory pattern as seen [here](https://github.com/BonusPlay/unnamed-sdvx-clone/commit/2e4a31ac34688a81c89b2687715d9e6aba0c71af#diff-375080b954ced9f59f3245e038db8f4f) and [here](https://github.com/BonusPlay/unnamed-sdvx-clone/blob/refactors/Graphics/include/Graphics/Shader.hpp).
This cleans up a lot of code and makes it very clear where logic is placed at. Also allows to write new implementations of current interfaces and switching to using them by replacing a single LoC. Since this project doesn't have to follow ABI at all I believe that migrating to factory functions and proper interfaces would improve project tremendously.

**Comment by Hinara** [comment]
Do you mean first compilation time or differential compilation ?
Some modules are not that bad like the audio module, but the core have a lot of functionalities, and modifying one file of it makes effectively recompile the whole project.
